src/Compilers/Core/Portable/Emit/DebugDocumentsBuilder.cs

```diff
@@ -47,6 +47,9 @@ internal void AddDebugDocument(Cci.DebugSourceDocument document)
             _debugDocuments.Add(document.Location, document);
         }
 
+        internal ImmutableArray<Cci.DebugSourceDocument> GetAllDebugDocuments()
+            => _debugDocuments.Values.ToImmutableArray();
```

Comment:
>_debugDocuments.Values.ToImmutableArray(); [](start = 15, length = 42)

This will result in non-deterministic PDB. We need to order the documents somehow.

Reply:
We could also optimize this to avoid allocations and extra work. Return IReadOnlyDictionary from this method. Enumerate the dictionary in AddRemainingDebugDocuments to filter out documents that are already indexed. Sort the remaining and add them.

---
In reply to: [333275492](https://github.com/dotnet/roslyn/pull/39136#discussion_r333275492) [](ancestors = 333275492)

Reply:
Sounds good. I just think that filtering is not necessary because we add only item not yet added.

---
In reply to: [333278515](https://github.com/dotnet/roslyn/pull/39136#discussion_r333278515) [](ancestors = 333278515,333275492)

Reply:
The filtering is an optimization, so that we don't need to sort all the documents upfront just to realize we only need to add a few.

Reply:
Let me try

---
In reply to: [333303087](https://github.com/dotnet/roslyn/pull/39136#discussion_r333303087) [](ancestors = 333303087)

Reply:
Thank you, @tmat! Updated.

---
In reply to: [333303897](https://github.com/dotnet/roslyn/pull/39136#discussion_r333303897) [](ancestors = 333303897,333303087)